mpc-core/src/garble/evaluator.rs

```diff
@@ -0,0 +1,108 @@
+use cipher::{consts::U16, BlockCipher, BlockEncrypt};
+
+use crate::{
+    block::{Block, SELECT_MASK},
+    garble::Error,
+};
+use mpc_circuits::{Circuit, Gate};
+
+use super::{
+    circuit::{prepare_inputs, BinaryLabel},
+    EncryptedGate, GarbledCircuit,
+};
+
+/// Evaluates AND gate
+#[inline]
+pub(crate) fn and_gate<C: BlockCipher<BlockSize = U16> + BlockEncrypt>(
+    cipher: &C,
+    x: &Block,
+    y: &Block,
+    encrypted_gate: &[Block; 2],
+    gid: usize,
+) -> Block {
+    let s_a = x.lsb();
+    let s_b = y.lsb();
+
+    let j = gid;
+    let k = gid + 1;
+
+    let hx = x.hash_tweak(cipher, j);
+    let hy = y.hash_tweak(cipher, k);
+
+    let w_g = hx ^ (encrypted_gate[0] & SELECT_MASK[s_a]);
+    let w_e = hy ^ (SELECT_MASK[s_b] & (encrypted_gate[1] ^ *x));
+
+    w_g ^ w_e
+}
+
+/// Evaluates XOR gate
+#[inline]
+pub(crate) fn xor_gate(x: &Block, y: &Block) -> Block {
+    *x ^ *y
+}
+
+pub fn evaluate<C: BlockCipher<BlockSize = U16> + BlockEncrypt>(
+    cipher: &C,
+    circ: &Circuit,
+    input_labels: &[BinaryLabel],
+    encrypted_gates: &[EncryptedGate],
+) -> Result<Vec<BinaryLabel>, Error> {
+    let mut labels: Vec<Option<Block>> = vec![None; circ.len()];
+
+    // Insert input labels
```

Comment:
I don't like the idea of the generator setting ids on the input labels he is sending. I think he should just send them as one blob and let the evaluator assign ids.

Not only we spend extra ~3% bandwidth on the ids, but also he may do some funny business by assigning malicious ids.
(I know that right now the attack will not go through. But it may slip in later when someone decides to refactor this part of the code).

Reply:
Can you elaborate on what the issue is with denoting the ids? If someone were to refactor this code they would be responsible to not introduce vulnerabilities. I can't think of how an attack would work. The generator is always capable of lying about the labels they send. The ids just provide some structure to help with catching bugs and code flow.
`prepare_inputs` does some checks on the input labels to make sure there is no funny business.

Reply:
> If someone were to refactor this code they would be responsible to not introduce vulnerabilities.
From my experience, it is safer just not to introduce pitfalls at all.
If you don't see eye-to-eye with me, that's ok, you can mark this as resolved.

Reply:
It's not that we don't see eye-to-eye, I'm just trying to understand the issue you're seeing. If the generator is sending what appears to be uniformly random input labels to the evaluator, what can he do to trick the evaluator?

Reply:
(Nothing with the current checks.)
But potentially he can set wrong ids and overwrite evaluator's labels with his.

Reply:
We can't guard against that in `fn evaluate`, we'll have to do it sometime between deserializing the message from the generator and here. Fortunately with Rust's typesystem we can ensure this. I will add a new type `SanitizedInputLabels` which forces the user to perform these checks before evaluating the circuit.